### DISCUSSION TITLE
Merge Confidence and OpenSSF

### DISCUSSION THREAD

**Discussion by adam-moss** [renovatebot/renovate#25125] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

latest official docker build

### If you're self-hosting Renovate, select which platform you are using.

gitlab.com

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

Merge confidence ratings combined with the OpenSSF scoring.

### What you tried so far.

Earlier today I enabled the `security:openssf-scorecard` preset (we already had the `mergeConfidence:all-badges` preset enabled.

From looking at a few merge requests however it appears these may be mutually exclusive?

Is that the case or am I misunderstanding something (again 😬 )

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

**Comment by adam-moss** [comment]
@rarkins apologies for the direct ping, could you offer some advice here?  Is it simply a case of defining my own preset that merges the two configs together or is there an alternative?

**Reply by rarkins** [reply]
Current PR body columns aren't mergeable, as that wouldn't let people _remove_ default columns. You could keep both presets but just define your own column list

**Reply by secustor** [reply]
I have thought some time ago if we should add a `addColumns` parameter similiar to `addLabels`. This would work here too and would make presets combineable.

**Reply by adam-moss** [reply]
Thank you both for the quick reply, I'll try defining the columns and let you know how I get on 😄

**Reply by adam-moss** [reply]
Hmm.  I'm struggling to get this to work, and am confused for another reason too 😭 

In my `config.json` I have

```json
{
    ...
    globalExtends: [
        ...
        "mergeConfidence:all-badges",
        "security:openssf-scorecard",
    ],
    ...
}
```

And have found 1 MR where all the columns are present:

<img width="1302" alt="Screenshot 2023-11-21 at 17 42 47" src="https://github.com/renovatebot/renovate/assets/2951486/563991fd-8a6e-4773-b351-c4bde067b957">


If I add 

```json
    prBodyColumns: [
        "Package",
        "Type",
        "Update",
        "Change",
        "Age",
        "Adoption",
        "Passing",
        "Confidence",
        "OpenSSF",
    ],
```

to my `config.json` it still generates the MR as if only `security:openssf-scorecard` were enabled which is what I'm seeing across the vast majority of MRs.

<img width="757" alt="Screenshot 2023-11-21 at 17 40 59" src="https://github.com/renovatebot/renovate/assets/2951486/80c61cb9-39af-4654-b03a-230d4c28cf7f">

**Reply by rarkins** [reply]
Preset columns are resolved later than config.js as they are in packageRules

**Reply by adam-moss** [reply]
Ok, got a bit further with this.

Adding a custom preset:

```json
{
    "$schema": "https://docs.renovatebot.com/renovate-schema.json",
    "description": "Merge the output of mergeConfidence:all-badges and security:openssf-scorecard.  See https://github.com/renovatebot/renovate/discussions/25125 for rationale.",
    "packageRules": [
        {
            "matchPackagePatterns": [".*"],
            "prBodyColumns": [
                "Package",
                "Type",
                "Update",
                "Change",
                "Age",
                "Adoption",
                "Passing",
                "Confidence",
                "OpenSSF"
            ]
        }
    ]
}
```

Gives what I was hoping for in _some_ circumstances.  In others however it is updating the existing MR to _remove_ dependency updates, notably pre-commit.

🤔

**Reply by adam-moss** [reply]
Right after a bit more debugging I'm pretty close:

**config.js**
```js
    ...
    extends: [
        ...,
        "gitlab>redact/redact/renovate//shared-configs/default-body-columns#fix/merge-confidence-openssf",
    ],
    ...
    prBodyDefinitions: {
        OpenSSF: "",
    },
    ...
```

**default-body-columns.json**
```json
{
    "$schema": "https://docs.renovatebot.com/renovate-schema.json",
    "description": "Merge the output of mergeConfidence:all-badges and security:openssf-scorecard.  See https://github.com/renovatebot/renovate/discussions/25125 for rationale.",
    "packageRules": [
        {
            "matchPackagePatterns": [".*"],
            "prBodyColumns": [
                "Package",
                "Type",
                "Update",
                "Change",
                "Age",
                "Adoption",
                "Passing",
                "Confidence",
                "OpenSSF"
            ]
        }
    ]
}
```

Gives me merge requests like:

<img width="1190" alt="Screenshot 2023-12-05 at 21 22 02" src="https://github.com/renovatebot/renovate/assets/2951486/6aaa2378-2186-4a04-9303-fac147118cd4">

<img width="1307" alt="Screenshot 2023-12-05 at 21 07 44" src="https://github.com/renovatebot/renovate/assets/2951486/3e175172-0f18-48f7-b5a6-d4d0e669eb66">


The "dummy" `prBodyDefinition` in `config.js` feels "dirty".  Is there a better approach you'd recommend?

Of the two example MRs:

- In the first I wouldn't expect to see an OpenSSF Scorecard value but is there a reason the column itself isn't rendered?
- In the second is there a way to stop the merge confidence attributes for displaying those broken links?

Finally OpenSSF Scorecard supports GitLab too, is it possible to update the preset to add that?  Currently it restricts itself to GitHub?

Thanks!

**Reply by HonkingGoose** [reply]
> The "dummy" `prBodyDefinition` in `config.js` feels "dirty". Is there a better approach you'd recommend?

I can't find a config option in the docs to _append_ new columns to the default `prBody` configuration. So looks like the dirty way is the only way, for now at least.

I'll let a maintainer respond to the rest of your questions. 😉

**Comment by xendren** [comment]
This would be helpful as a feature request.  It seems a merging of the configs could work here?  I'm just getting into using Renovate, so I am not sure if there is a reason that isn't possible.  The extends and presets are extremely nice features, but this seems to be a gap.  Is there a reason that merge-confidence wouldn't include openSSF column by default?  It seems pertinent to be included in that setting to me.  Devs would want to know all of that information to have confidence in merging.

**Reply by rarkins** [reply]
It should be possible to configure both Merge Confidence columns AND the OSSF column at the same time (beware of the width required though). If for some reason it's not practical to do so, let's have a reproduction repo and we can work out a way to make it possible. There's no need for the Merge Confidence preset itself to include OSSF.

**Reply by xendren** [reply]
Do you mean the workaround that is mentioned above?  Create a new repo to manage custom presets, and add the dummy prBodyDefinition and the extends to the config?

**Reply by rarkins** [reply]
I mean we should try to make it easier and a workaround unnecessary
